Add docs describing saved tensor hooks

Varal7 opened the pull request:
Add section to the Autograd mechanics docs to describe the recently
exposed saved tensors (#52451), how to register packing / unpacking
hooks (#60975) and how to use default hooks (#61834)

Sister PR: #62361 (will add a link from autograd.rst to notes/autograd in whatever PR does not land first)

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -36,6 +36,51 @@ flow statements, that can change the overall shape and size of the graph at
+When defining a custom Python :class:`~torch.autograd.Function`, you can use
+:func:`~torch.autograd.function._ContextMethodMixin.save_for_backward` to save
+tensors during the forward pass and
+:attr:`~torch.autograd.function.NestedIOFunction.saved_tensors` to retrieve them
```
That shouldn't point to the `NestedIOFunction` version here rigth?

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -36,6 +36,51 @@ flow statements, that can change the overall shape and size of the graph at
+    print(x is y.grad_fn._saved_self) # True
+
+
+In the previous code, ``y.grad_fn._saved_self`` shares the same storage as `x`.
```
Actually they do share the same storage. And they always do, except for inplace ops that save their input (that is being modified inplace and so obviously we need another Tensor with different storage to store the original value).
Here you want to say that they might not be the same Tensor object.

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -598,3 +643,83 @@ chain rule:
+You can control :ref:`how saved tensors are packed / unpacked
+<saved-tensors-doc>` by defining a pair of ``pack_hook`` / ``unpack_hook`` hooks.
+The ``pack_hook`` function should take a tensor as its single argument but can
+return any python object (e.g. another tensor, a tuple, or even a filename).
```
nit: the third element in your or is not a type. Maybe "or even a string containing a filename"?

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -598,3 +643,83 @@ chain rule:
+The ``pack_hook`` function should take a tensor as its single argument but can
+return any python object (e.g. another tensor, a tuple, or even a filename).
+The ``unpack_hook`` function takes as its single argument the output of
+``pack_hook`` and should return the original tensor to be used in the backward
```
"the original tensor" -> we should define that better here.
It can actually be a different Tensor object. It can have a different storage, etc

The requirement is that it should be a Tensor that has the same content.
Any autograd-related metadata should be ignored in particular as they will be overwritten.

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -598,3 +643,83 @@ chain rule:
+.. warning::
+
+    Both functions should avoid performing inplace operations on the input or
+    undefined behavior may arise.
```
I am not a fan of have a UB here.
Why is inplace during packing forbidden?

Inplace during unpacking, we should say that this is forbidden and this function should be callable multiple times and should return the same data everytime.

Varal7 replied:
For example:

```python
def pack(t):
    t += 2
    return t - 2

torch.autograd.graph.set_saved_tensors_default_hooks(pack, lambda x: x)
a = torch.ones(5, requires_grad=True)
print(a)  # tensor([1., 1., 1., 1., 1.], requires_grad=True)
y = a * a
print(a)  # tensor([5., 5., 5., 5., 5.], requires_grad=True)
```

I guess that's a case where we can actually catch the inplace modification. The question is whether we prefer making inplace operations during pack forbidden or UB.

albanD replied:
It's not UB here. It actually does modify the input inplace. But it doesn't lead to any silently wrong gradients beyond that.
But I do agree that the user shouldn't do that  and we want to forbid it :D

Varal7 replied:
I'd argue that in the correct form, it's an UB because the way the backward formula was implemented can be subject to change without us notifying the user, as it is an implementation detail that they should not rely on. 
I can make a PR to always throw an error when the tensor is modified and make it very well defined behavior :)

Varal7 replied:
Here's a concrete example, powered by the fact that different backends have different derivative formulas for a given op.

```python
def f(device):
  lstm = nn.LSTM(5, hidden_size=5, num_layers=2, bidirectional=True, batch_first=True).to(device)
  for x in lstm.parameters():
    x.data.zero_()
  x = torch.zeros(5, 5, 5).to(device)
  return lstm(x)[0].sum()

print(f("cpu"))   # tensor(0., grad_fn=<SumBackward0>)
print(f("cuda"))  # tensor(0., device='cuda:0', grad_fn=<SumBackward0>)

def inc(x):
  x += 1
  return x - 1

torch.autograd.graph.set_saved_tensors_default_hooks(inc, lambda x: x)
print(f("cpu"))   # tensor(2700., grad_fn=<SumBackward0>)
print(f("cuda"))  # tensor(497.9154, device='cuda:0', grad_fn=<SumBackward0>)
torch.autograd.graph.reset_saved_tensors_default_hooks()
```

Collab: https://colab.research.google.com/drive/1XD9D2ack7WIAyrUYujuW0oKm8VYLkqAF?usp=sharing

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -598,3 +643,83 @@ chain rule:
+.. code::
+
+    def pack_hook(x):
+        name = os.path.join(tmp_dir, str(uuid.uuid4()))
```
I like this example, but we should use it to show how to handle the lifetime as well I think.
In particular, with this code, you're gonna "leak" everything you write on disk.

I think you want to show that the file on disk should live as long as the object your returned. And that the unpack must NOT delete the file as it might be called multiple times.

Varal7 replied:
The solution I have in mind is to clear the tmp_dir after each backward. Is there a better way you were thinking?

albanD replied:
You can have a file handle here that will delete the file when the handle is deleted. You can check the python's TempFile, I think they do that.

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -598,3 +643,83 @@ chain rule:
+
+    If you maintain a reference to a :class:`SavedTensor` after the saved
+    tensors have been released (i.e. after backward has been called) and call
+    its :meth:`~torch.autograd.SavedTensor.register_hooks`, undefined behavior
```
nit: mention explicitly that this is forbidden I think. And then say that if the user still does it, it will be UB as we cannot catch that call all the time.

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -598,3 +643,83 @@ chain rule:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Alternatively, you can register a pair of default hooks which will be applied
+to all saved tensors that are created in that context.
```
Do you want to give an example of this here?

Also maybe show how the hook can filter so that it only moves to disk Tensors that have more than "threshold" elements in them.

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -598,3 +643,83 @@ chain rule:
+to all saved tensors that are created in that context.
+
+Note that using a pair of identity hooks is not a no-op as it forces
+packing / unpacking even in cases where Pytorch would have chosen to not
```
nit: I would phrase this slightly differently. More like having a pair of hooks disables all the optimization in place to reduce Tensor object creation.

albanD reviewed (commented):
Looks quite good!
Just some small comments.

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
+
+Some operations need intermediary results to be saved during the forward pass
+in order to execute the backward pass. For example, the function
+:math:`x\mapsto x^2` needs to save the input :math:`x` because it is needed to
+compute the gradient.
```
Nit: it could technically save `x^2` so I wouldn't say it "needs to save the input". Maybe just "the function saves the input x to compute the gradient"

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
+:attr:`~torch.autograd.function.Function.saved_tensors` to retrieve them
+during the backward pass. See :doc:`/notes/extending` for more information.
+
+For operations that Pytorch defines (e.g. ``aten::pow``), tensors are
```
nit: Pytorch -> PyTorch.

Also should we talk about `torch.pow` instead of `aten::pow`?

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
+    x = torch.randn(5, requires_grad=True)
+    y = x.pow(2)
+    print(x.equal(y.grad_fn._saved_self)) # True
+    print(x is y.grad_fn._saved_self) # True
```
Nit: Python style guide says two spaces have to exist before the start of an inline comment

zou3519 commented on docs/source/notes/autograd.rst:
```diff
@@ -36,6 +36,52 @@ flow statements, that can change the overall shape and size of the graph at
+For operations that Pytorch defines (e.g. ``aten::pow``), tensors are
+automatically saved as they are needed. You can explore which tensors are saved
+by a certain ``grad_fn`` by looking for its attributes starting with the prefix
+``_saved``.
```
I assume we prefixed this with `_` because we don't want users to rely on it. Is this true? If so we should explain that these are for debugging only

Varal7 committed:
Add docs describing saved tensor hooks

Add section to the Autograd mechanics docs to describe the recently
exposed saved tensors (#52451), how to register packing / unpacking
hooks (#60975) and how to use default hooks (#61834)

ghstack-source-id: 9d9aa297afb84df0257f7a4a82a98952e4eb2549
Pull Request resolved: https://github.com/pytorch/pytorch/pull/63647

albanD commented on docs/source/notes/autograd.rst:
```diff
@@ -36,6 +36,57 @@ flow statements, that can change the overall shape and size of the graph at
+:math:`x\mapsto x^2` saves the input :math:`x` to compute the gradient.
+
+When defining a custom Python :class:`~torch.autograd.Function`, you can use
+:func:`~torch.autograd.function._ContextMethodMixin.save_for_backward` to save
```
cc @soulitzer this link will need updating when we rename that class

albanD reviewed (approved):
LGTM!
Thanks for the update following the latest PRs
